sources/filesystem/gitpythonfs/gitpythonfs/core.py

```diff
@@ -0,0 +1,210 @@
+                and/or ref can be passed in the URL instead of arguments. (The
+                actual file path should not contain "@" or ":"). Examples:
+                    When instantiating GitPythonFileSystem:
+                    /some_folder/my_repo
+                    When calling open(), open_files() etc:
+                    gitpythonfs:///some_folder/my_repo:path/to/intro.md
+                    gitpythonfs:///some_folder/my_repo:mybranch@path/to/intro.md
+            ref (str): (To be implemented). A branch, tag or commit hash to use.
+                Defaults to head of the local repo.
+        """
+        super().__init__(**kwargs)
+        self.repo_path = path
+        self.repo = git.Repo(self.repo_path)
+
+    @classmethod
+    def _strip_protocol(cls, path: str) -> str:
+        path = super()._strip_protocol(path).lstrip("/")
+        if ":" in path:
+            path = path.split(":", 1)[1]
+        if "@" in path:
+            path = path.split("@", 1)[1]
+        return path.lstrip("/")
+
+    # ToDo support arguments in url, like this example from git fsspec implementation:
+    @staticmethod
+    def _get_kwargs_from_urls(path: str) -> Dict[str, str]:
+        if path.startswith("gitpythonfs://"):
+            path = path[14:]
+        out = {}
+        if ":" in path:
+            out["path"], path = path.split(":", 1)
+        if "@" in path:
+            out["ref"], path = path.split("@", 1)
+        return out
+
+    def _git_type_to_file_type(self, object: git.Object) -> str:
+        if isinstance(object, git.Blob):
+            return "file"
+        elif isinstance(object, git.Tree):
+            return "directory"
+        else:
+            msg = f"There is no fileystem object type corresponding to Git object type: {type(object).__name__}"
+            raise TypeError(msg)
+
+    def _details(
+        self, object: git.Object, include_committed_date: bool = True
+    ) -> Dict[str, Union[str, int]]:
+        """
+        Retrieves the details of a Git object.
+
+        Args:
+            object (git.Object): The Git object to retrieve details for.
+            include_committed_date (bool, optional): Whether to include the committed date. Defaults to True.
+                Getting the committed date is an expensive operation and will slow down
+                walk(), a method that is extensively used by fsspec for find(), glob() etc.
+
+        Returns:
+            dict: A dictionary containing the details typical for fsspec.
+        """
+        # commit=next(self.repo.iter_commits(paths=object.path, max_count=1))
+        details = {
+            "name": object.path,
+            "type": self._git_type_to_file_type(object),
+            "mime_type": object.mime_type if isinstance(object, git.Blob) else None,
+            "size": object.size,
+            "hexsha": object.hexsha,
+            # "committed_date": commit.committed_date,
+        }
+
+        if include_committed_date:
```

Comment:
yeah this will create a process for each file to run git rev parse or something similar. this is indeed a no go... what we should do is to index and cache a whole tree or a path when doing `ls` with a single git command.

ie. look here https://stackoverflow.com/questions/1964470/whats-the-equivalent-of-subversions-use-commit-times-for-git

`git --no-pager whatchanged --pretty=%at`
this is really fast

then here you'll just look into cache

Reply:
First cut of this is at 9daaf9a
It is not as fast as I had hoped.

- I used `git log --raw` instead of `git whatchanged`. `log` has better documentation.  Apparently they both use `git rev-list` behind the scenes.  Also using a git pathspec to limit what is returned to the scope of what `ls()` is doing.
- The caching is very conservative - limited the single tree (not recursive) that `ls()` is operating on.  So one call to `git log --raw` for every folder in the repo. 
- Overall the pattern is more pre-fetching rather than caching.
- Could do one `git log --raw` for the entire folder hierarchy, for given ref.   It starts to push the architecture of fsspec though.  As far as I can see, we would need to override (or examine at runtime) something up the hierarchy (or call stack) - like glob(), walk() - to detect when they're about to deeply traverse the repo. The glob pattern could then be applied in the pathspec of the git command so it only pre-fetches revisions that will be used.
- Note complexity that any cache needs to be indexed by ref

II will try to find where it is slow.

Reply:
2nd cut.  Eagerly caches all repo revisions with one `git log --raw`.   Results for `walk(path="", ref=""HEAD)` per repo:
- dlt: 0.5s
- wine: ~50s  (has 168k commits, 10k files)

Is this acceptable?